### PR TITLE
Fields in viscosity operators

### PR DESCRIPTION
**Pull request by simone-silvestri**
this PR allows to have a dynamic viscosity and diffusivity

if a location is specified through the keyword argument `loc` the function will be interpolated automatically, otherwise the function will have `lx, ly, lz` in the signature
```
using Oceananigans.Operators: ζ₃ᶠᶠᶜ, div_xyᶜᶜᶜ, ℑxyᶜᶜᵃ, Azᶜᶜᶜ, Δx, Δy

C₄ = 0.2

@inline function νhb_1(i, j, k, grid, clock, fields, C₄) 

    δ  = div_xyᶜᶜᶜ(i, j, k, grid, fields.u, fields.v)
    ζ  = ℑxyᶜᶜᵃ(i, j, k, grid, ζ₃ᶠᶠᶜ, fields.u, fields.v)
    
    Δ₂ = Azᶜᶜᶜ(i, j, k, grid)
    return C₄ * Δ₂^2 sqrt(δ^2 + ζ^2)
end

@inline function νhb_2(i, j, k, grid, lx, ly, lz, clock, fields, λ) 
    Δ = (1 / (1 / Δx(i, j, k, grid, lx, ly, lz)^2 + 1 / Δy(i, j, k, grid, lx, ly, lz)^2) )^2
    return Δ / λ
end

biharmonic_viscosity_1  = HorizontalScalarBiharmonicDiffusivity(ν=νhb_1, discrete_form=true, loc = (Center, Center, Center), parameters = C₄)
biharmonic_viscosity_2  = HorizontalScalarBiharmonicDiffusivity(ν=νhb_2, discrete_form=true, parameters = 1days)
```

### CODE REVIEW

**Comment by simone-silvestri** [comment]
Or maybe we can do this just with `AbstractOperations`, the problem is that we have to define a "helper" model

File: src/TurbulenceClosures/abstract_scalar_biharmonic_diffusivity_closure.jl
Comment (glwagner) [inline src/TurbulenceClosures/abstract_scalar_biharmonic_diffusivity_closure.jl:37]: why `clk, F` rather than `F, clk`?

File: src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl:46]: This change (and all the other changes associated with it) are not necessary to achieve the aim of this PR, right?

**Comment by glwagner** [comment]
> Or maybe we can do this just with `AbstractOperations`, the problem is that we have to define a "helper" model

I think extending the capabilities of the `DiscreteDiffusionFunction` is the right approach here.

But I also think we should only make the necessary changes in abstract_scalar_diffusivity.jl and abstract_biharmonic_scalar_diffusivity.jl and leave the tendency kernels and existing closures (like AMD) unchanged. The only thing that needs to be done is `merge(U, C)` there.

File: src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl:40]: This will fail for `PrescribedVelocityFields`?

**Comment by glwagner** [comment]
Also, having this capability would be awesome!

**Comment by glwagner** [comment]
Some of the stuff being implemented here was discussed in #2580

File: src/TurbulenceClosures/abstract_scalar_diffusivity_closure.jl
Comment (glwagner) [inline src/TurbulenceClosures/abstract_scalar_diffusivity_closure.jl:86]: I think it's better to be explicit here for purposes of self-documentation (if we can)... using `args...` has led to some pain in the past

File: src/Operators/interpolation_utils.jl
Comment (glwagner) [inline src/Operators/interpolation_utils.jl:118]: This has the same name as 

https://github.com/CliMA/Oceananigans.jl/blob/499c2eb8199877c9226e4c7d8a63e88616364972/src/Fields/interpolate.jl#L88

File: src/TurbulenceClosures/discrete_diffusion_function.jl
Comment (glwagner) [inline src/TurbulenceClosures/discrete_diffusion_function.jl:69]: `getdiffusivity`?

File: src/Advection/upwind_biased_fifth_order.jl
Comment (glwagner) [inline src/Advection/upwind_biased_fifth_order.jl:10]: Why are you changing the definition of the boundary buffer in this PR?

**Review by glwagner — CHANGES_REQUESTED** [review]

File: src/Advection/upwind_biased_fifth_order.jl
Comment (glwagner) [reply src/Advection/upwind_biased_fifth_order.jl:10]: The number "2" is the number of buffer cells over which we limit to second order advection. Therefore `CenteredSecondOrder` has a buffer of "0" by definition.

We can change the definition of this number if we want, but some rationale might be helpful...

File: src/TurbulenceClosures/abstract_scalar_biharmonic_diffusivity_closure.jl
Comment (tomchor) [inline src/TurbulenceClosures/abstract_scalar_biharmonic_diffusivity_closure.jl:26]: Can we be a little more verbose with the arguments and write `fields` instead of `F` when this function's called? (Assuming `F` indeed passes the fields.)

I think this helps developers less familiar with the code (myself very much included)

File: src/TurbulenceClosures/turbulence_closure_implementations/isopycnal_skew_symmetric_diffusivity.jl
Comment (tomchor) [inline src/TurbulenceClosures/turbulence_closure_implementations/isopycnal_skew_symmetric_diffusivity.jl:188]: Same comment here with `c`. Can we write out what `c` is in the calls to `diffusive_flux` to make it easier to understand? Is it tracers?

**Review by tomchor — CHANGES_REQUESTED** [review]
This is looking really good, @simone-silvestri!

I just posted a couple of comments that I think make the code easier to understand in general.

Can you also make sure @glwagner's comments are addressed before merging? Maybe mark them as "resolved" if you've already dealt with them

**Review by francispoulin — COMMENTED** [review]
I looked it over and it looks good! I will leave it up to others who know the closure schemes better than I do to approve this.

**Pull request merged**
